### PR TITLE
✨ PROS 4: Motor_Groups and Motor Inheritance

### LINKED ISSUE #430: PROS 4: Make MotorGroups inherit from Motor

### LINKED ISSUE #474: Create get_raw_positions method for motor groups
#### Expected Behavior:
A `get_raw_positions()` method that returns a `std::vector<std::int32_t>` of the raw encoder positions of a motor group

#### Actual Behavior:
The method does not exist

#### Additional Information
Check other motor methods that may need motor group implementations e.g. `get_temperatures()`, `get_powers()`, etc.

### PR DESCRIPTION
**Pull request by noam987**
Mostly hardware tested, but not cleaned up, or correct docs, but ready for review.
Clean up and docs will be separate PR as I want to be able to use this functional code  in other PRs and features as well.

#### Summary:
Adds motor groups that directly call C API. 
Motors inherit from Motor_Groups in the C++ API

#### Motivation:
Motors inheriting from motor groups makes the C++ API much easier to maintain as if any adjustment needs to be made to the C++ API for the physical motor device, it only needs to be implemented for motor groups instead of both motors and motor groups.
Allows for more dynamic motor groups.

##### References (optional):
closes #474 
closes #430 (Ended up having motors inherit from motor_group)
closes 

#### Test Plan:
Call every method in Motor_Group and Motor.

### CODE REVIEW

File: src/devices/vdml_motors.cpp [hunk 1]
```diff
@@ -17,265 +17,662 @@ namespace pros {
...
+	}
+	return return_vector;
 }
 
-std::int32_t Motor::get_current_draw(void) const {
+pros::v5::Motor_Brake Motor_Group::get_brake_mode(void) const {
+	empty_motor_group_check(pros::v5::Motor_Brake::invalid);
 	push_motor_configuration();
-	return motor_get_current_draw(_port);
+	return static_cast<pros::v5::Motor_Brake>(motor_get_brake_mode(_ports[0]));
```
Comment (djava) [inline src/devices/vdml_motors.cpp:211]: you can just call the `Motor_Brake` ctor instead of casting right?

Comment (djava) [reply src/devices/vdml_motors.cpp:211]: applies below as well

File: src/devices/vdml_motors.cpp [hunk 2]
```diff
@@ -17,265 +17,662 @@ namespace pros {
...
+
+void Motor_Group::erase_port(std::int8_t port) {
+	int i = 0;
+	while (i < _ports.size()) {
+		if (std::abs(_ports[i]) == port) {
+			_ports.erase(_ports.begin() + i);
+		} else {
+			i++;
+		}
+	}
```
Comment (djava) [inline src/devices/vdml_motors.cpp:675]: use `std::erase` + `std::remove_if` instead.

File: src/devices/vdml_motors.cpp [hunk 3]
```diff
@@ -288,24 +685,47 @@ std::ostream& operator<<(std::ostream& os, const pros::Motor& motor) {
...
 
-void Motor::push_motor_configuration() const {
-	_motor_mutex.take();
+void Motor_Group::push_motor_configuration() const {
+	_motor_group_mutex.take();
 	set_gearing(_gearset);
 	set_reversed(_reverse);
 	set_encoder_units(_encoder_units);
-	_motor_mutex.give();
+	_motor_group_mutex.give();
```
Comment (djava) [inline src/devices/vdml_motors.cpp:703]: `std::lock_guard`

**Review by djava — CHANGES_REQUESTED** [review]
This is a good start to the `Motor_Group`/`Motor` refactor, but I think there are a lot of things that  should be rethought or redesigned.

I'm... not so sold on using `Motor_Group` as the base class with the intention of polymorphic usage.  I think this idea kind of got away from us here..
- The original idea was `AbstractMotor` so that libraries could use `AbstractMotor` to accept either `Motor` or `Motor_Group` without too much more thought about it.
- The idea was changed to having `Motor_Group` inherit from `Motor` to avoid user confusion about what the heck an `AbstractMotor` is.
  - So now libraries could take `Motor&` which would be understandable for users and still allow the polymorphism.
  - I wasn't a huge fan of how badly this broke traditional ideas about class hierarchy (or the philosophy shift toward prioritizing beginner users), but was kind of willing to accept that the API was pretty clean from the user-side, and that's what's important after all.
- Now, `Motor` inherits from `Motor_Group` instead? I think this loses on all sides.
  - It seems confusing to both beginner and advanced users as to libraries would be taking `Motor_Group&` as the only option, since its not obvious that that it's using polymorphism.
  - It breaks traditional class hierarchy, the kernel-side code organization is confusing, and the lowest-level interchangeability of a `Motor` with a one-element `Motor_Group` is strange at best.
  - Its a huge performance pessimization, as every motor construction now constructs a `std::vector` which needs to heap allocate to store the single integer.  This removes the triviality of `Motor` construction, which was one of the very cool things about how they were designed in PROS 3.

In my opinion, the class hierarchy should be reconsidered to either one of the first two ideas (preferably the first imo), or disregarded entirely.  I consider this design as it stands a sizable downgrade from PROS 3.

File: src/devices/vdml_motors.c [hunk 4]
```diff
@@ -24,9 +24,11 @@
 #define MOTOR_MOVE_RANGE 127
 #define MOTOR_VOLTAGE_RANGE 12000
 
+
 // Movement functions
 
-int32_t motor_move(uint8_t port, int32_t voltage) {
+int32_t motor_move(int8_t port, int32_t voltage) {
+	port = abs((int8_t) port);
```
[the remark this answers is not in the document]
Comment (WillXuCodes) [reply src/devices/vdml_motors.c:31]: oops missed a spot lol

[the remark this answers is not in the document]
Comment (WillXuCodes) [reply src/devices/vdml_motors.c:31]: > huh? we're just instantly discarding the sign of the port? i thought we were conditionally reversing based on it...? applies to almost all functions below as well

was handled on the C++ end but yeah it's kinda sketch.

File: include/pros/motors.hpp [hunk 5]
```diff
@@ -129,23 +129,41 @@ class Motor {
...
-	explicit Motor(const std::uint8_t port, const pros::Color gearset_color);
+	explicit Motor_Group(std::vector<std::int8_t> ports, const Motor_Gears gearset);
 
-	explicit Motor(const std::uint8_t port, const bool reverse);
+	explicit Motor_Group(std::vector<std::int8_t> ports, const pros::Color gearset_color);
 
-	explicit Motor(const std::uint8_t port);
+	explicit Motor_Group(std::vector<std::int8_t> ports, const bool reverse);
+
+	explicit Motor_Group(std::vector<std::int8_t> ports);
```
[the remark this answers is not in the document]
Comment (WillXuCodes) [reply include/pros/motors.hpp:167]: > I also think that accepting `pros::Color` instead of `Motor_Gears` is wholly unnecessary at best and very unclear at worst. Accepting only a couple members of an enum (I assume that's whats happening here?) feels like terrible practice to me.

This was a community request that I don't think is best practice either, but with respect to those who want it it's not hard to support either.

File: src/devices/vdml_motors.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (djava) [reply src/devices/vdml_motors.c]: So the C API doesn't support negative ports reversing? That is pretty sketch. I'd probably expect that behavior to be consistent across the APIs.

[the remark this answers is not in the document]
Comment (WillXuCodes) [reply src/devices/vdml_motors.c]: > So the C API doesn't support negative ports reversing? That is pretty sketch. I'd probably expect that behavior to be consistent across the APIs.

yep, well this is something we could look into in the future. Right now our priority is just getting something out and seeing what other critical features may be broken from a functional perspective.

File: src/devices/vdml_motors.cpp [hunk 6]
```diff
@@ -215,8 +215,8 @@ std::vector<pros::v5::Motor_Brake> Motor_Group::get_every_brake_mode(void) const
 	std::vector<pros::v5::Motor_Brake> return_vector;
 	empty_motor_group_check_vector(pros::v5::Motor_Brake::invalid, return_vector);
 	push_motor_configuration();
-	for (int i = 0; i < _ports.size(); i++) {
-		return_vector.emplace_back(static_cast<pros::v5::Motor_Brake>(motor_get_brake_mode(_ports[i])));
+	for (auto it = _ports.begin() + 1; it < _ports.end(); it++) {
```
Comment (djava) [inline src/devices/vdml_motors.cpp:218]: the +1 in the initializer must be removed here, it's only for if you're explicitly performing the action on the first element outside the loop, like in `brake()`.  Same applies to most of the functions below

Comment (noam987) [reply src/devices/vdml_motors.cpp:218]: Good catch. I missed this when I used replace all at midnight

File: src/devices/vdml_motors.cpp [hunk 7]
```diff
@@ -17,265 +17,662 @@ namespace pros {
...
+	return motor_modify_profiled_velocity(_ports[0], velocity);
 }
 
-double Motor::get_actual_velocity(void) const {
+double Motor_Group::get_actual_velocity(void) const {
+	empty_motor_group_check(PROS_ERR_F);
 	push_motor_configuration();
-	return motor_get_actual_velocity(_port);
+	return motor_get_actual_velocity(_ports[0]);
 }
```
[the remark this answers is not in the document]
Comment (Richard-Stump) [reply src/devices/vdml_motors.cpp:196]: The reason we are going for the design we did is that we thought it would be more confusing to users if we had a base class for motors. Instead we opted for the idea that motors are a "specialization" of the motor groups. This removes one extra class in the API, and means that we don't have to explain the abstract motor to those new to C++. 

I see your point about the single motor getters/setters. Maybe we could make them take an optional index with 0 as the default? Then again, that is kinda redundant with the operator[] method.

[the remark this answers is not in the document]
Comment (Richard-Stump) [reply src/devices/vdml_motors.cpp:196]: Additional thought about my 2nd paragraph: users can use the "every" methods and index them for specific motors.

File: src/devices/vdml_motors.cpp (hunk 3 above)
Comment (Richard-Stump) [reply src/devices/vdml_motors.cpp:705]: I think we should stick with the pros mutexes since this is in the kernel's api.

File: include/pros/motors.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Richard-Stump) [reply include/pros/motors.hpp:167]: > Surely this long list of constructors could be greatly reduced to 3 or 4 using default parameters instead of all these overloads?

I think that using optional arguments would be good to shorten the number of constructors; however, motors do have defaults for many things without us calling VexOS. Making a bunch of the arguments optional would require us to have some redundant/inconsistent default values, and/or more calls to the VexOS. I think it's fine how it is.

File: include/pros/motors.hpp [hunk 8]
```diff
@@ -756,80 +783,22 @@ class Motor {
...
 	 *   while (true) {
 	 *     motor = master.get_analog(E_CONTROLLER_ANALOG_LEFT_Y);
-	 *     std::cout << "Is the motor stopped?: " << motor.is_stopped();
+	 *     std::cout << "Motor Voltage: " << motor.get_voltage();
 	 *     pros::delay(2);
 	 *   }
 	 * }
 	 * \endcode
 	 */
-	virtual std::int32_t is_stopped(void) const;
```
Comment (Richard-Stump) [inline include/pros/motors.hpp:832]: Was this removal intentional?

File: src/devices/vdml_motors.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (djava) [reply src/devices/vdml_motors.cpp:196]: > Instead we opted for the idea that motors are a "specialization" of the motor groups
see top level code review comment for my thoughts on this

> I see your point about the single motor getters/setters. Maybe we could make them take an optional index with 0 as the default?
maybe, but then its weird for the `Motor` objects that it takes a param. This whole hierarchy doesn't really work with how OOP hierarchy is meant to be.

> users can use the "every" methods and index them for specific motors
This is very bad for performance, every call creates a vector which needs to heap allocate which is expensive.  In reality, probably wont matter for 99% of users, but..

File: include/pros/motors.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (djava) [reply include/pros/motors.hpp:167]: > Making a bunch of the arguments optional would require us to have some redundant/inconsistent default values, and/or more calls to the VexOS

I don't think this is actually a big deal, but perhaps 
```cpp
explicit Motor_Group(std::initializer_list<std::int8_t> ports, 
                     const std::optional<Motor_Gears> gearset = std::nullopt,
                     const std::optional<bool> reverse = std::nullopt,
                     const std::optional<Motor_Units> encoder_units = std::nullopt);
```
if you're really concerned about it? you would just call that as 
```cpp
Motor_Group({8,1,2}, gearset = Motor_Gears::Blue, reverse = false);
```
so users dont actually need to worry about the `std::optional`. I guess leaving it as is could be fine too though.

File: src/devices/vdml_motors.cpp (hunk 3 above)
Comment (djava) [reply src/devices/vdml_motors.cpp:705]: Why? `std::lock_guard` is just a RAII mutex wrapper, it's not an actual mutex. It does the exact same thing as calling `pros::Mutex::take()` and `::give()` but with less potential for mistakes.

**Comment by Richard-Stump** [comment]
@djava 
> I'm... not so sold on using Motor_Group as the base class with the intention of polymorphic usage. I think this idea kind of got away from us here..

We appreciate the feedback; however, we are sticking with this design for motor groups.

File: src/devices/vdml_motors.cpp (hunk 2 above)
Comment (Richard-Stump) [reply src/devices/vdml_motors.cpp:675]: This probably would be a better solution here.

File: include/pros/motors.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Richard-Stump) [reply include/pros/motors.hpp:167]: This would be a nice solution if we decide to change it. I will think about this.
